zkevm-circuits/src/circuit_tools/cell_manager.rs

```diff
@@ -60,77 +94,199 @@ impl<F: Field> Cell<F> {
 
 impl<F: Field> Expr<F> for Cell<F> {
     fn expr(&self) -> Expression<F> {
-        self.expression.as_ref().unwrap().clone()
+        self.expression.unwrap().clone()
     }
 }
 
 impl<F: Field> Expr<F> for &Cell<F> {
     fn expr(&self) -> Expression<F> {
-        self.expression.as_ref().unwrap().clone()
+        self.expression.unwrap().clone()
+    }
+}
+
+// #[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+// pub(crate) enum CellType {
+//     StoragePhase1,
+//     StoragePhase2,
+//     StoragePermutation,
+//     LookupByte,
+//     Lookup(Table),
+// }
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub(crate) enum CellType_<T: CustomTable> {
+    StoragePhase1,
+    StoragePhase2,
+    StoragePermutation,
+    LookupByte,
+    Lookup(T),
+}
+
+pub trait CustomTable: Clone + Copy + Debug + PartialEq + Eq + PartialOrd + Ord + Hash{
+    fn matches_to(&self, other: &Self) -> bool;
```

Comment:
True, in fact `CustomTable` is just a marker trait because we need a type parameter for generic tables. It kind of make sense to me seeing specific tools bounded by specific table and config:
```
pub struct ConstraintBuilder<F, T: CustomTable> {
    constraints: Vec<(&'static str, Expression<F>)>,
    max_degree: usize,
    conditions: Vec<Expression<F>>,
    pub lookups: Vec<LookupData<F>>,
    pub lookup_tables: Vec<LookupData<F>>,
    pub cell_manager: Option<CellManager<F, T>>,
}

pub struct CellManager<F, T: CustomTable> {
    pub(crate) width: usize,
    pub(crate) height: usize,
    pub(crate) cells: Vec<Cell<F>>,
    pub(crate) columns: Vec<CellColumn<F, T>>,
    pub(crate) phase_config: PhaseConfig<T>,
    pub(crate) copy_columns: usize,
}

pub struct PhaseConfig<T> {
    phase3: Vec<(T, usize)>, //lookup
    phase2: usize, // rlc
    phase1: usize, // byte
}
```
Maybe we can also move[ these functions](https://github.com/CeciliaZ030/zkevm-circuits/blob/a04940ce8acc00559ca920d33f4bd10de25c5866/zkevm-circuits/src/table.rs#L34) in `CustomTable` as well.

Reply:
Or maybe the opposite, allow `LookupTable` to be used in the cellmanager as trait instead of `CustomTable`?